ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+"""
+Midterm review for cs 114
+
+Instructor: Cameron Morland - cjmorland@uwaterloo.ca
+
+This review will cover materials from module 1 to 6, from all the slides 
+and the textbook materials
+
+"""
+
+"""
+Let's start with the very basic
```

Comment:
A note about the types int, float, and str. At some level of the implementation, your program will allocate memory for each instance of any object. 

A string is an array of chars (characters), and each ASCII char is eight bits. Something worth noting is that while there is no formal definition for a byte, every modern CPU uses the convention that eight bits is a byte. 

There are many different types of integers as far as the C programming language is concerned. The first way ints are classified is by their bit size. There are 8, 16 (short), 32 (long), and 64-bit (long long) integers. The number of values that these ints can take on is given by 2 raised to the power of their bit size. So an 8-bit integer can take on 256 values. The actual values they take on is determined by the other way ints are differentiated, _unsigned_ and _signed_. They will always include zero. An unsigned int is always non-negative. So an unsigned 8-bit integer can take on any integer from 0 to 255. A signed 8-bit int can take on values from -128 to 127. 

For floats, they can be 32-bit, 64-bit (double), and 128-bit (long double), although the latter is seldom used and has no IEEE standard for its behaviour. The bit size determines the decimal point precision. A 32-bit int in C guarantees up to 7 decimal points of precision. A 64-bit int can store up to 14 decimal points. However, in Python, a 64 bit integer can have 16 digits of precision. 

Being dynamically typed, the developer doesn't have any control on what bit-size is used for each type at implementation when using native Python types. However, this can create several issues. The first is that Python often defaults to a bit-size that more than big enough for the task, which results in more memory being allocated than necessary, which also slows down the program ever so slightly when reading from memory. The other issue is interfacing with binary streams with other programs that are often written in C or C++. In fact, since the developer has no control over native Python object bit sizes, writing to binary in general can be very difficult since knowing the bit sizes of objects is required. To get around this, you can use the data types in the Numpy library, which are just the C types exposed to the Python environment.